python/hail/experimental/ldscore.py

```diff
@@ -71,107 +69,88 @@ def ld_score(entry_expr, annotation_exprs,
-        *[(ht_annotations.annotate(annotation=hl.str(x),
-                                   value=hl.float(ht_annotations[x]))
-                         .select('annotation', 'value'))
-          for x in annotation_names]
-    )
-    mt_annotations = ht_annotations.to_matrix_table(row_key=variant_key,
-                                                    col_key=['annotation'])
-
-    cols = mt_annotations['annotation'].collect()
-    col_idxs = {i: cols[i] for i in range(len(cols))}
-
-    G = BlockMatrix.from_entry_expr(entry_expr)
-    A = BlockMatrix.from_entry_expr(mt_annotations.value)
-
-    n = G.n_cols
-
-    R2 = ((G @ G.T) / n) ** 2
-    R2_adj = R2 - (1.0 - R2) / (n - 2.0)
-
-    positions = [(x[0], float(x[1])) for x in
-                 hl.array([mt.locus.contig, hl.str(position_expr)]).collect()]
-    n_positions = len(positions)
-
-    starts = np.zeros(n_positions, dtype='int')
-    stops = np.zeros(n_positions, dtype='int')
-
-    contig = '0'
-    for i, (c, p) in enumerate(positions):
-        if c != contig:
-            j = i
-            k = i
-            contig = c
-
-        min_val = p - window_size
-        max_val = p + window_size
-
-        while j < n_positions and positions[j][1] < min_val:
-            j += 1
-
-        starts[i] = j
-
-        if k == n_positions:
-            stops[i] = k
-            continue
-
-        while positions[k][0] == contig and positions[k][1] <= max_val:
-            k += 1
-            if k == n_positions:
-                break
-
-        stops[i] = k
-
-    R2_adj_sparse = R2_adj.sparsify_row_intervals([int(x) for x in starts],
-                                                  [int(x) for x in stops])
-    L2 = R2_adj_sparse @ A
-
-    tmp_bm_path = new_temp_file()
-    tmp_tsv_path = new_temp_file()
-
-    L2.write(tmp_bm_path, force_row_major=True)
-    BlockMatrix.export(tmp_bm_path, tmp_tsv_path)
-
-    ht_scores = hl.import_table(tmp_tsv_path, no_header=True, impute=True)
+    N = mt.count_cols()
+
+    R2 = hl.row_correlation(entry_expr, block_size) ** 2
+    R2_adj = (N - 1.0)/(N - 2.0) * R2 - 1.0/(N - 2.0)
+
+    starts, stops = hl.linalg.utils.locus_windows(locus_expr,
+                                                  radius,
+                                                  coord_expr)
+    R2_adj_sparse = R2_adj.sparsify_row_intervals(starts, stops)
+
+    tmp_R2_adj_sparse_file = new_temp_file()
+    R2_adj_sparse.write(tmp_R2_adj_sparse_file, force_row_major=True)
+    R2_adj_sparse = BlockMatrix.read(tmp_R2_adj_sparse_file)
+
+    if not annotation_exprs:
+        cols = ['univariate']
+        col_idxs = {0: 'univariate'}
+        L2 = R2_adj_sparse.sum(axis=1)
+    else:
+        ht = mt.select_rows(*wrap_to_list(annotation_exprs)).rows()
+        ht = ht.annotate(univariate=hl.literal(1.0))
+        names = [x for x in ht.row if x not in ht.key]
+
+        ht_union = hl.Table.union(
+            *[(ht.annotate(name=hl.str(x),
```

Comment:
I'm confused, can you explain what's going on here? Is the challenge to turn row annotations into a block matrix, so you have to melt the table to create the matrix table to create the block matrix? That'll be quite bad performance-wise if there are tons of annotations, so it means we're missing some useful primitives, like turning a table into a matrix table when the fields other than the key are of the same type.

Reply:
Yes, the challenge is going from a table to a block matrix. I didn't see any other way using existing functions than this melting approach. I agree it is not ideal.